homework-submission/W2-homework/todo-app/src/App.js

```diff
@@ -0,0 +1,59 @@
+import React, { Component } from 'react';
+import './App.css';
+import TodoItems from './Components/TodoItems/TodoItems';
+import AddItem from './Components/AddItem/AddItem';
+
+/**
+ * This class component is the container of the Todo List application
+ * @return Here is the return of the application
+ */
+class App extends Component {
+  state = {
+    todoItems: [
+      {
+        id: 1,
+        description: 'Get out of bed',
+        deadline: '2017-09-11',
+        done: true,
+      },
+      {
+        id: 2,
+        description: 'Brush teeth',
+        deadline: '2017-09-10',
+        done: false,
+      },
+      {
+        id: 3,
+        description: 'Eat breakfast',
+        deadline: '2017-09-09',
+        done: false,
+      },
+    ],
+  };
+
+  deleteTodoItem = id => {
+    let todoItems = this.state.todoItems.filter(todoItem => {
+      return todoItem.id !== id;
+    });
+    this.setState({ todoItems });
+  };
+
+  addTodoItem = todoItem => {
+    let todoItems = this.state.todoItems;
+    todoItem.id = Math.random() * 10 + 1;
+    todoItems.push(todoItem);
```

Comment:
You are manipulating the state directly!

If you tried the following:
```javascript
    console.log(this.state.todoItems.length) // if this 3
    todoItems.push(todoItem)
    console.log(this.state.todoItems.length) // this is changed to be 4, which it should not do that
```

the `.setState` method is on top of a complete lifecycle on behalf of eyes which made react makers make it asynchronous, and that is why the strange behavior you have saw here.

This is an important thing to keep in mind avoiding Bugs in the future, and understanding how react dealing with it's own state.

you could solve this as the following allowing your self to add an item without touching the state directly:
```javascript
this.setState({
  todoItems: [
    ...this.state.todoItems,
    {
      ...todoItem,
      id: /* Some unique id... */
    }
  ]
})
```

Reply:
Many thanks for this advice.